src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,242 @@ end
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+# Iterator over values of grouping columns for each group
+function _groupvalues(gd::GroupedDataFrame)
+    T = Tuple{_grouptypes(gd)...}
+    return (T(gd.parent[gd.idx[s], gd.cols]) for s in gd.starts)
+end
+
+# Get values of grouping columns for single group by its index
+_groupvalues(gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
+
+# Get value of grouping column by group and column index
+function _groupvalue(gd::GroupedDataFrame, groupidx::Integer, colidx::Integer)
+    return gd.parent[gd.idx[gd.starts[groupidx]], gd.cols[colidx]]
+end
+
+
+"""
+    GroupKey
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{N, T<:Tuple}
+    parent::GroupedDataFrame
+    idx::Int
+
+    function GroupKey{N,T}(parent, idx) where {N,T}
+        if length(N) != length(T.parameters)  # correct way to get length of T?
+            throw(ErrorException("Length of names and field types do not match"))
+        end
+        return new{N,T}(parent, idx)
+    end
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey{N}) where N = length(N)
+Base.names(::GroupKey{N}) where N = N
+Base.keys(key::GroupKey) = names(key)
+Base.values(key::GroupKey{N,T}) where {N,T} = _groupvalues(parent(key), getfield(key, :idx))::T
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
+Base.iterate(key::GroupKey, i::Integer) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = _groupvalue(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey{N}, n::Symbol) where N
+    i = findfirst(==(n), N)
+    i === nothing && throw(KeyError(n))
+    return key[i]
+end
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
+    catch e
+        throw(ErrorException("$(typeof(key)) has no property $p"))
+    end
+end
+
+Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
+
+Base.NamedTuple(key::GroupKey{N}) where {N} = NamedTuple{N}(values(key))
```

Comment:
Looking at this line what is the benefit of your definition of `GroupKey` over simply defining it as:
```
struct GroupKey{T<:GroupedDataFrame}
    parent::T
    idx::Int
end
```

Reply:
Pretty much just the type inference you get from the extra type parameters.

Reply:
The issue is (as commented below) if we later really get this type-inference benefit (so we pay a cost during creation to have type stability but the question is if we use this type stability later to get the benefits of paying this cost here).

As an extreme case have a look at e.g. https://discourse.julialang.org/t/bad-performance-of-group-by-of-dataframes/30061 from today that refers to `by` - in that use case it is actually better to have type unstable code.

I am not saying that what you do is bad - but we should just be sure that it is superior to what I proposed in a typical usage scenario as both have their pros and cons. Thank you for looking into this!